Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c

```diff
@@ -1,6 +1,63 @@
 #include "Training.h"
 
-void trainingTask(uint8_t* data)
+void trainingTask (uint8_t* data)
 {
+	validData = 0;
+	int motorOneState = motorCheck (*data);
+	int motorTwoState = motorCheck (*(data + 1));
+	int lightState = lightsCheck (*(data + 2));
+	
+	if (motorOneState && motorOneState == motorTwoState) // set outputArray values of motor if motors data are valid, motors are in sync
+	{
+		outputArray[0] = *data;
+		outputArray[1] = *(data + 1);
+		validData = validData | (1 << FIRSTMOTOR);  // set first motor bit in validData to 1 if motor's data is valid
+		validData = validData | (1 << SECONDMOTOR); // set second motor bit in validData to 1 if motor's data is valid
+	}
+	else
+	{
+		outputArray[0] = 0;
+		outputArray[1] = 0;
+	}
+	
+	if (lightState) // set outputArray values of lights if lights data are valid
+	{
+		outputArray[2] = *(data + 2);
+		validData = validData | (1 << LIGHTSBOARD);  // set lights bit in validData to 1 if lights data is valid
+	}
+	else
+	{
+		outputArray[2] = 0;
+	}	
+}
+
+int motorCheck (uint8_t testMotor)
+{
+	uint8_t motorMask = 0b11000001;
+	
+	if ((testMotor & motorMask) == 0b10000001) // check the case that motor is in reverse (power on, negative velocity and reverse)
+		return 1;
+	else if ((testMotor & motorMask) == 0b11000000) // check the case that motor is in forward (power on, positive velocity and forward)
+		return 2;
+	else if ((testMotor & 0b10000001) == 0b00000001) // check case that power is off and motor on "forward"
+		return 3;
+	else if (testMotor == 0) // check case that power is off
+		return 4;	
```

Comment:
One thing I just remembered. It's possible that the velocity is 0 while the car is on and has a direction. Unfortunately, we've chosen sign-mag for representing the velocity, so 0 has 2 representations: 0b100000 and 0b000000. So it's technically not incorrect for the direction of the car to be forward, and the sign of the velocity to be 1 **if** the magnitude of the velocity is 0.

Another point I forgot to point out is that if the motors are off, the velocity should be 0. Then again you can argue that the motors could've recently been on and just turned off so because of inertia, the velocity isn't 0. I'm fine with  it either way, but you'll just have to explicitly state the latter assumption.

Reply:
Can the direction also be reverse while the power is off? If so I can make one if statement to check this statement and that v=0 when the motor is off.

Reply:
Yes, that's also valid